### ISSUE TITLE
Feature request: support beeswarm's "compactswarm"

### ISSUE THREAD

**Issue by bersbersbers** [csdaw/ggbeeswarm2#1]
See https://github.com/aroneklund/beeswarm/commit/ef9b318244677408cc9995769fd80cc65fffbe50 and later commits.

**Comment by csdaw** [comment]
Hi @bersbersbers,

I've implemented the `compactswarm` method now. Let me know if it works for you. Also just to note, eventually the changes made in this fork will be incorporated into the actual ggbeeswarm package and this fork will be abandoned.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(ggplot2)

set.seed(123)
distro <- list(runif = runif(100, min = -3, max = 3), 
               rnorm = rnorm(100))

distro2 <- as.data.frame(rev(distro)) %>% 
  tidyr::pivot_longer(
    everything(), 
    values_to = "value", 
    names_to = "variable"
  )

distro2 %>% 
  ggplot(aes(x = rev(variable), y = value)) + 
  ggbeeswarm2::geom_beeswarm(method = "swarm", spacing = 0.8)
```

![](https://i.imgur.com/bCbFKIY.png)

``` r
distro2 %>% 
  ggplot(aes(x = rev(variable), y = value)) + 
  ggbeeswarm2::geom_beeswarm(method = "compactswarm", spacing = 0.8)
```

![](https://i.imgur.com/J91NoJX.png)

<sup>Created on 2021-07-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by bersbersbers** [comment]
Thanks, `compactswarm` vs `swarm` looks really sensible.

I do have issues converting my `ggbeeswarm` calls to `ggbeeswarm2`, though. I am still playing with `size` and `spacing`, but some (not all) `ggbeeswarm2` plots are just hugely V-shaped (`swarm`) or circular (`compactswarm`). But I guess this is a different issue, so I will probably revisit `compactswarm` when merged into `ggbeeswarm`.

**Comment by bersbersbers** [comment]
Regarding this other problem, I think it is nicely illustrated here:

```R
do_plot <- function(filename, fun) {
    png(
        type = "cairo",
        filename = filename,
        width = 1.9,
        height = 1.9,
        units = "in",
        res = 720,
    )
    plot(ggplot2::ggplot(data.frame(), ggplot2::aes(x = 0, y = 1:10)) +
        fun())
    dev.off()
}
do_plot("ggbeeswarm.png", ggbeeswarm::geom_beeswarm)
do_plot("ggbeeswarm2.png", ggbeeswarm2::geom_beeswarm)
```

![image](https://user-images.githubusercontent.com/12128514/126978824-dbd58055-7433-4782-b7cd-eed46d6251f5.png)

**Comment by csdaw** [comment]
After a bit of digging, I can explain why this is happening (see below) and it is sort of a bug and sort of not. In any case I'll make a change to `PositionBeeswarm` which will make it (hopefully) behave a bit more sensibly.

---

So the swarming in both `ggbeeswarm` and `ggbeeswarm2` relies on `swarmx()` from the `beeswarm` package. The `beeswarm()` base plotting function from the `beeswarm` package also relies on `swarmx()`. 

If I make a plot using `beeswarm()` and the same parameters for `png()` you will see what happens:

```r
df <- data.frame(
  x = rep(0, 10),
  y = 1:10
)

png(
  type = "cairo",
  filename = "beeswarm.png",
  width = 1.9,
  height = 1.9,
  units = "in",
  res = 720
)
beeswarm::beeswarm(y ~ x, data = df, m = "swarm")
dev.off()
```

<img src="https://user-images.githubusercontent.com/45983385/127313938-6c53998e-0da2-4994-95da-b9430c36394c.png" width=50% height=50%>

`swarmx()` 'senses' the points are overlapping and performs the swarming as requested. The reason for the different behaviour in `ggbeeswarm2` vs `ggbeeswarm` is down to how they estimate the size of the points on the plot and hence if they think the points are overlapping or not.

`ggbeeswarm2` approximates the way that `beeswarm` determines the size of the points using `graphics::xinch()` and `graphics::yinch()` which change relative to the size of the plotting device (`graphics::par("pin")`) and the plot region (`graphics::par("usr")`). What I didn't realise was that `graphics::par("pin")` is influenced by `graphics::par("mar")` which can lead to odd results as you have shown, so I'll need to push a commit to account for this.

On the other hand `ggbeeswarm` uses a simplified method and relies only on the x and y ranges of the data, divided by a 'magic number'.

If I take a slightly modified version of your example, the result of this difference in implementation is made obvious.

```r
do_plot <- function(filename, fun) {
  png(
    type = "cairo",
    filename = filename,
    width = 2.3, # increase plot device size a bit
    height = 2.3, # increase plot device size a bit
    units = "in",
    res = 720,
  )
  # plot 100 points between 1 and 10
  set.seed(1)
  plot(ggplot2::ggplot(data.frame(), ggplot2::aes(x = 0, y = runif(100, 1, 10))) +
         fun())
  dev.off()
}
do_plot("ggbeeswarm.png", ggbeeswarm::geom_beeswarm)
do_plot("ggbeeswarm2.png", ggbeeswarm2::geom_beeswarm)
```



<p float="left">
<img src="https://user-images.githubusercontent.com/45983385/127341238-25dd28d3-1899-4550-99d3-118b27f5091e.png" width=45% height=45%>
<img src="https://user-images.githubusercontent.com/45983385/127341249-9b79200a-dd10-4fc8-a51a-9341082d7d89.png" width=45% height=45%>
</p>

**Comment by bersbersbers** [comment]
Thanks, this is definitely getting better!

Unfortunately, I think my previous example was too simplistic, as I am seeing the same issues as before in my complete example. This slightly more complex example (namely, including an outlier) still behaves better in `ggbeeswarm` than `ggbeeswarm2`:

```r
y <- 1:100
y <- c(y, -2000) # add outlier
do_plot <- function(filename, fun) {
    png(
        type = "cairo",
        filename = filename,
        width = 1.9,
        height = 1.9,
        units = "in",
        res = 720,
    )
    plot(ggplot2::ggplot(data.frame(), ggplot2::aes(x = 0, y = y)) +
        fun(size=1/1000))
    dev.off()
}
do_plot("ggbeeswarm.png", ggbeeswarm::geom_beeswarm)
do_plot("ggbeeswarm2.png", ggbeeswarm2::geom_beeswarm)
```

![image](https://user-images.githubusercontent.com/12128514/127616363-72dda2c3-a80d-4380-ba16-45aabbf20521.png)

Also, I noticed that `fun(size=1/3)` or `fun(size=3)` does not seem to influence the *distribution* of points at all, am I right there? Is that maybe part of the issue above? Do I need to play with `spacing` when I change `size`, given that `cex` is unavailable in `ggbeeswarm2`?

**Comment by csdaw** [comment]
Yes you're correct, you need to play with `spacing` when you adjust the size of the points with `size` when using `ggbeeswarm2`. In `ggbeeswarm` when you change the point `size`, the distribution also changes by itself because of (I believe) `ggplot2::standardise_aes_names()`.

`spacing` is just what I renamed the `cex` argument when I made this fork (which was originally intended for personal use only) and so this behaviour is lost. When the changes here are added to `ggbeeswarm` the `cex` argument will retain its name for backward compatibility/to avoid annoying a lot of people 😂. 

But thanks for bringing this issue up, I'll need to propagate the fix to `ggbeeswarm`. In any case, here is what happens when you play around with spacing:

```r
y <- c(1:100, -2000)

do_plot <- function(filename, fun) {
  png(
    type = "cairo",
    filename = filename,
    width = 1.9,
    height = 1.9,
    units = "in",
    res = 720,
  )
  plot(ggplot2::ggplot(data.frame(), ggplot2::aes(x = 0, y = y)) +
         fun(size = 1/1000, spacing = 0.2))
  dev.off()
}
do_plot("ggbeeswarm.png", ggbeeswarm::geom_beeswarm)
do_plot("ggbeeswarm2.png", ggbeeswarm2::geom_beeswarm)
```

<p float="left">
<img src="https://user-images.githubusercontent.com/45983385/127635381-fdd6e07e-61ab-4e77-9cf9-ceabf5f27ce9.png" width=45% height=45%>
<img src="https://user-images.githubusercontent.com/45983385/127635387-8dc2fd95-f5cd-4b66-95ff-8bb693880c2e.png" width=45% height=45%>
</p>

**Comment by bersbersbers** [comment]
Thanks again! Here's one more (the final?) issue I came across:

```R
y <- c(
    0.0, 0.0, 0.0, 0.1, 0.1, 0.1, 0.2, 0.2, 0.2, 0.3, 0.3, 0.3, 0.4, 0.4, 0.4,
    0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.5, 0.5, 0.5, 0.6, 0.6, 0.6, 0.6, 0.6,
    0.6, 0.6, 0.6, 0.6, 0.7, 0.7, 0.7, 0.8, 0.8, 0.9
)
y <- c(y, -10)
do_plot <- function(filename, fun, ...) {
    png(
        type = "cairo",
        filename = filename,
        width = 5,
        height = 1,
        units = "in",
        res = 800,
    )
    plot(ggplot2::ggplot(data.frame(), ggplot2::aes(x = 0, y = y)) +
        ggplot2::coord_cartesian(ylim = c(0, 1)) +
        fun(...)) +
        NULL
    dev.off()
}
do_plot("ggbeeswarm2_swarm.png", ggbeeswarm2::geom_beeswarm, method = "swarm")
do_plot("ggbeeswarm2_compactswarm.png", ggbeeswarm2::geom_beeswarm, method = "compactswarm")
```

Is it expected that the outlier has such a big influence on the overall result? Is that specific due to my using `ylim`? Note that I am not changing `size` here in any way, and I wonder why the middle (that is, `(x,y) = (0, 0.5)`) is left empty in both variants.

![image](https://user-images.githubusercontent.com/12128514/127661770-bfb2a864-3f76-49bf-8aff-d6385e89fc22.png)

I suspect that the outlier increases the *computational* y limits, even though the visual y limits are much smaller. So computationally, the points are squished much closer vertically, so that `beeeswarm` thinks that the center is still occupied by the samples in `(0, 0)`. This may be something that can be fixed using spacing, but that will increase the horizontal spacing, which is not really necessary here. A proper solution will probably take into account the *visual* y limits.

**Comment by bersbersbers** [comment]
I suspect that 
https://github.com/csdaw/ggbeeswarm2/blob/dddff8952e56cec1f44481ebbd202e5087f58b1d/R/position-beeswarm.R#L113
and
https://github.com/csdaw/ggbeeswarm2/blob/dddff8952e56cec1f44481ebbd202e5087f58b1d/R/position-beeswarm.R#L117
are related to this - these are getting the limits of the scale, not the coordinate system.

**Comment by csdaw** [comment]
Yep you're right, the swarming for both `ggbeeswarm` and `ggbeeswarm2` rely on the scale rather than the coordinate system. Also something to note is both `ggbeeswarm` and `ggbeeswarm2` tend to give better results with a discrete scale (i.e. `x = as.factor(0)`.

In this case the swarming occurs within `ggbeeswarm2::PositionBeeswarm`, then after that the changes made to the coordinate system are applied within `ggplot2::GeomPoint`.

To take into account _visual_ y limits would require writing a custom `GeomBeeswarm`, which I honestly don't really would be worth the effort as it's probably better to just filter outliers prior to plotting.

So I think I'll leave things there.

This discussion has been very useful though, thank you! Now I've written the original `ggbeeswarm` swarming method back into this package as `method = "swarm"`. My altered swarming method is now `method = "swarm2"`. This will be very important for my PR to `ggbeeswarm` as it will preserve backwards compatibility.

Hopefully now calls to `ggbeeswarm2::geom_beeswarm(method = "swarm")` will perform identically to `ggbeeswarm::geom_beeswarm()`

**Comment by bersbersbers** [comment]
> it's probably better to just filter outliers prior to plotting

I am not sure I agree with this general statement, based on my very limited understanding of `ggplot` (separating data from graphics instructions). Then again, I can see why one might like to obtain the same swarm irregardless of different visual limits (e.g., to zoom into a specific region of a swarm plot shown in full in another panel). So I guess there is little we can do other than manually tweaking swarming parameters here and there when a plot nears its final state (which we already do for shape sizes etc, so this is not adding much extra complication anyway).

Thanks again!
